crates/executor/src/swarm.rs

```diff
@@ -0,0 +1,137 @@
+use futures::StreamExt;
+use libp2p::gossipsub::{self, IdentTopic};
+use libp2p::identity::Keypair;
+use libp2p::swarm::{NetworkBehaviour, SwarmEvent};
+use libp2p::{mdns, noise, tcp, yamux, SwarmBuilder};
+use std::error::Error;
+use std::time::Duration;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tracing::{debug, error};
+use zetina_common::graceful_shutdown::shutdown_signal;
+
+#[derive(NetworkBehaviour)]
+pub struct PeerBehaviour {
+    gossipsub: gossipsub::Behaviour,
+    mdns: mdns::tokio::Behaviour,
+}
+
+pub struct SwarmRunner {
+    handle: Option<JoinHandle<()>>,
+}
+
+impl SwarmRunner {
+    pub fn new(
+        p2p_local_keypair: &Keypair,
+        subscribe_topics: Vec<IdentTopic>,
+        mut transmit_topics: Vec<(IdentTopic, mpsc::Receiver<Vec<u8>>)>,
+        swarm_events_tx: mpsc::Sender<gossipsub::Event>,
+    ) -> Result<Self, Box<dyn Error>> {
+        let mdns = mdns::tokio::Behaviour::new(
+            mdns::Config::default(),
+            p2p_local_keypair.public().to_peer_id(),
+        )?;
+        let gossipsub = Self::init_gossip(p2p_local_keypair)?;
+        let behaviour = PeerBehaviour { gossipsub, mdns };
+        let local_keypair = p2p_local_keypair.clone();
+        let mut swarm = SwarmBuilder::with_existing_identity(local_keypair)
+            .with_tokio()
+            .with_tcp(
+                tcp::Config::default().port_reuse(true),
+                noise::Config::new,
+                yamux::Config::default,
+            )?
+            .with_quic()
+            .with_behaviour(|_| behaviour)?
+            .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(60)))
+            .build();
+
+        for topic in subscribe_topics {
+            swarm.behaviour_mut().gossipsub.subscribe(&topic)?;
+        }
+
+        swarm.listen_on("/ip4/0.0.0.0/udp/5680/quic-v1".parse()?)?;
+        swarm.listen_on("/ip4/0.0.0.0/tcp/5681".parse()?)?;
```

Comment:
smol suggestion, same context with [::]: above
```suggestion
        // Listen on both IPv4 and IPv6 for UDP/QUIC
        swarm.listen_on("/ip4/0.0.0.0/udp/5680/quic-v1".parse()?)?;
        swarm.listen_on("/ip6/::/udp/5680/quic-v1".parse()?)?;

        // Listen on both IPv4 and IPv6 for TCP
        swarm.listen_on("/ip4/0.0.0.0/tcp/5681".parse()?)?;
        swarm.listen_on("/ip6/::/tcp/5681".parse()?)?;
```

Reply:
I have one more comment on this, it seems like it should have port 
5678 and 5679 no? doesn't match with docker file and delegator swarm